File: cublas/dot.c
#include <stdio.h>
#include <math.h>
#include <cuda_runtime.h>
#include <cublas_v2.h>

int main(void)
{
    int i;
    const int n = 100;
    double x[n], y[n];
    double *x_, *y_, *z_;
    double z_ref = 0.0;
    double z;
    cublasHandle_t handle;

    // initialise data and calculate reference value on CPU
    for (i=0; i < n; i++) {
        x[i] = sin(i) * 2.3;
        y[i] = cos(i) * 1.1;
        z_ref += x[i] * y[i];
    }
    cublasCreate(&handle);

    // use GPU pointers to keep data on GPU
    cublasSetPointerMode(handle, CUBLAS_POINTER_MODE_DEVICE);

    // allocate + copy initial values
    cudaMalloc((void **)(&x_), sizeof(double) * n);
    cudaMalloc((void **)(&y_), sizeof(double) * n);
    cudaMalloc((void **)(&z_), sizeof(double));
    cudaMemcpy(x_, x, sizeof(double) * n, cudaMemcpyHostToDevice);
    cudaMemcpy(y_, y, sizeof(double) * n, cudaMemcpyHostToDevice);

    // calculate dot on GPU
    cublasDdot(handle, n, x_, 1, y_, 1, z_);

    // copy result back to host and print with reference
    cudaMemcpy(&z, z_, sizeof(double), cudaMemcpyDeviceToHost);
    printf("reference: %f\n   result: %f\n", z_ref, z);

    cublasDestroy(handle);
    return 0;
}

File: cuda-numpy/dot.py
from math import sin, cos

from numpy import array

from _dot import dot
from _mem import Dallocate, deallocate, Dmemcpy_h2d, Dpeek

blocks = 32
threads = 256
n = 10000

# initialise data and calculate reference values on CPU
x = array([sin(i) * 2.3 for i in range(n)], float)
y = array([cos(i) * 1.1 for i in range(n)], float)
z_ref = sum(x * y)

# allocate + copy initial values
x_ = Dallocate(n)
y_ = Dallocate(n)
z_ = Dallocate(1)
Dmemcpy_h2d(x_, x, n)
Dmemcpy_h2d(y_, y, n)

# calculate dot product on GPU (and reduce partial sums)
dot(blocks, threads, n, x_, y_, z_)

# copy result back to host and print with reference
z = Dpeek(z_)
print(' reference: {0}'.format(z_ref))
print('    result: {0}'.format(z))
